python/ray/client_builder.py

```diff
@@ -57,7 +65,95 @@ def connect(self) -> ClientInfo:
+        self.create_cluster_without_searching = (address == "local")
+        self._init_args = {}
+        self._server = None
+
+    def resources(
+            self,
+            *,
+            num_cpus: Optional[int] = None,
+            num_gpus: Optional[int] = None,
+            resources: Optional[Dict[str, str]] = None,
+            object_store_memory: Optional[int] = None) -> "ClientBuilder":
+        self._init_args["num_cpus"] = num_cpus
+        self._init_args["num_gpus"] = num_gpus
+        self._init_args["resources"] = resources
+        self._init_args["object_store_memory"] = object_store_memory
+        return self
+
+    def dashboard(self,
+                  *,
+                  include: bool = True,
+                  host: str = "127.0.0.1",
+                  port: Optional[int] = None) -> "ClientBuilder":
+        assert self.create_cluster_without_searching
+        self._init_args["include_dashboard"] = include
+        self._init_args["dashboard_host"] = host
+        self._init_args["dashboard_port"] = port
+        return self
+
+    def _create_local_cluster(self):
+        def ray_connect_handler(job_config=None):
+            with disable_client_hook():
+                if not ray.is_initialized():
+                    ray.init(job_config=job_config, **self._init_args)
+
+        port, _ = get_unused_port()
+        self.address = f"localhost:{port}"
+        logger.info(f"Creating a server on {self.address}")
+        self._server = client_server.serve(self.address, ray_connect_handler)
+        atexit.register(ray.shutdown, True)
+        atexit.register(self._server.grpc_server.stop, 0)
+
+    def _find_or_create_cluster(self):
+        """
+        Searches for a local cluster, and looks in redis for the port
+        of Ray Client Server. If this fails, a cluster is manually launched.
+        """
+        redis_addresses = find_redis_address()
+        candidates = set()
+        # NOTE If multiple Ray clusters are running locally (i.e. some launched
+        # via ray.client("local")), but only one has the Client Server address
+        # available in Redis, we will connect to that one.
+        for address in redis_addresses:
+            redis_client = create_redis_client(address, REDIS_DEFAULT_PASSWORD)
+            port = redis_client.get(RAY_CLIENT_SERVER_PORT_REDIS_KEY)
+            if port is not None:
+                candidates.add("localhost:" + port.decode("UTF-8"))
+        if len(candidates) == 0:
+            logger.warning(
+                f"Found clusters at: {redis_addresses}, but was unable to "
+                "determine the Ray Client Server Port in Redis. "
+                "Starting a new cluster instead.")
+            self._create_local_cluster()
+        elif len(candidates) == 1:
+            self.address = candidates.pop()
+        else:
+            raise ConnectionError(
+                f"Found multiple Ray Client Servers: {candidates}.\n")
```

Comment:
The term `Client Server` should not appear in anything user facing, why not say "Found multiple local Ray clusters"?